pkg/handlers/middleware.go

```diff
@@ -44,3 +54,130 @@ func RequireValidLicenseIDMiddleware(next http.Handler) http.Handler {
 		next.ServeHTTP(w, r)
 	})
 }
+
+// Code for the cache middleware
+type CacheEntry struct {
+	RequestBody  []byte
+	ResponseBody []byte
+	StatusCode   int
+	Expiry       time.Time
+}
+
+type cache struct {
+	store map[string]CacheEntry
+	mu    sync.RWMutex
+}
+
+func NewCache() *cache {
+	return &cache{
+		store: map[string]CacheEntry{},
+	}
+}
+
+func (c *cache) Get(key string) (CacheEntry, bool) {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+
+	entry, found := c.store[key]
+	if !found || time.Now().After(entry.Expiry) {
+		return CacheEntry{}, false
+	}
+	return entry, true
+}
+
+func (c *cache) Set(key string, entry CacheEntry, duration time.Duration) {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	// Clean up expired entries
+	for k, v := range c.store {
+		if time.Now().After(v.Expiry) {
+			delete(c.store, k)
+		}
+	}
+
+	entry.Expiry = time.Now().Add(duration)
+	c.store[key] = entry
+}
+
+type responseRecorder struct {
+	http.ResponseWriter
+	Body       *bytes.Buffer
+	StatusCode int
+}
+
+func (r *responseRecorder) WriteHeader(code int) {
+	r.StatusCode = code
+	r.ResponseWriter.WriteHeader(code)
+}
+
+func (r *responseRecorder) Write(b []byte) (int, error) {
+	r.Body.Write(b)
+	return r.ResponseWriter.Write(b)
+}
+
+func CacheMiddleware(next http.HandlerFunc, duration time.Duration) http.HandlerFunc {
+	// Each handler has its own cache to reduce contention for the in-memory store
+	cache := NewCache()
+
+	return func(w http.ResponseWriter, r *http.Request) {
+		body, err := io.ReadAll(r.Body)
+		if err != nil {
+			logger.Error(errors.Wrap(err, "cache middleware - failed to read request body"))
+			http.Error(w, "cache middleware: unable to read request body", http.StatusInternalServerError)
+			return
+		}
+
+		r.Body = io.NopCloser(bytes.NewBuffer(body))
+
+		hash := sha256.Sum256([]byte(r.Method + "::" + r.URL.Path))
```

Comment:
what about query parameters, if we are making this generic?

Reply:
and why is body removed from key now?

Reply:
I am open to making it more generic.

Reply:
Done: https://github.com/replicatedhq/replicated-sdk/pull/193/commits/28bc4e75fbf64ecc539daf6c6e88f138194fc7ed 

I also have to move towards `reflect.DeepEqual` since `maps.Equals` only compares primitives.